docs/codeql/codeql-cli/about-codeql-workspaces.rst

```diff
@@ -0,0 +1,84 @@
+
+In most cases, the CodeQL workspace and all CodeQL packs contained in it should be stored in the same git repository so the development environment is more easily sharable.
+
+The ``codeql-workspae.yml`` file
+--------------------------------
+
+A CodeQL workspace is defined by a ``codeql-workspace.yml`` yaml file. This file contains a ``provide`` block, and optionally an ``ignore`` block. The ``provide`` block contains a list of glob patterns that define the CodeQL packs that are available in the workspace. The ``ignore`` block contains a list of glob patterns that define CodeQL packs that are not available in the workspace. Each entry in the ``provide`` or ``ignore`` section must map to a path to a ``qlpack.yml`` file. All glob patterns are relative to the directory containing the workspace file. See `@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>`__ for a list of patterns accepted in this file.
+
+For example, the following ``codeql-workspace.yml`` file defines a workspace that contains all CodeQL packs recursively found in the ``codeql-packs`` directory, except for the packs in the ``experimental`` directory:
+
+.. code-block:: yaml
+
+    provide:
+      - "*/codeql-packs/**/qlpack.yml"
+    ignore:
+      - "*/codeql-packs/**/experimental/**/qlpack.yml"
+
+To verify that you have the correct ``codeql-workspace.yml`` file, run ``codeql pack ls`` command in the same directory as your workspace. The result of the command is a list of all CodeQL packs in the workspace.
+
+
+CodeQL workspaces and query resolution
+--------------------------------------
+
+All CodeQL packs in a workspace are available as source dependencies for each other when running any CodeQL command that resolves queries or packs. For example, when ``codeql pack install`` is run in a pack directory in a workspace, any dependency found in the workspace will not be downloaded to the package cache, nor will it be added to the resulting ``codeql-pack.lock.yml`` file. See `:ref:Adding and Installing Dependencies <adding-and-installing-dependencies>` for more information.
+
+Similarly, publishing a CodeQL query pack to the GitHub container registry using  ``codeql pack publish`` will always use dependencies found in the workspace instead of using dependencies found in the local package cache.
+
+This ensures that any local change to a query library in a dependency in the same workspace will be automatically reflected in the published query pack.
+
+.. pull-quote::
+
+  Note
+
+  Source dependencies are CodeQL packs that are resolved from the filesystem. They might be in the same CodeQL workspace, or specified as a path option in the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and version constraints are ignored. This ensures that during local development version mismatches can be ignored.
```

Comment:
```suggestion
  Source dependencies are CodeQL packs that are resolved from the local file system. These dependencies can be in the same CodeQL workspace, or specified as a path option using the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and any version constraints defined in the cache are ignored. This ensures that when you are developing CodeQL locally any version mismatches can be ignored.
```

I've suggested what I think is a simpler construction above, but it's true to say that I'm left wondering if that means that you're likely to publish a CodeQL pack with version mismatches that might cause your users problems.

Reply:
Thanks for the suggestion. It's not quite right. And I can understand where your confusion is coming from.  Version constraints are defined in the qlpack.yml. 

Source dependencies are generally useful in two situations:

- The main use case is when you are doing local development and want to override some dependencies as you work out some changes across multiple packs during development.
- You want to override a dependency in your cache for some reason (maybe there's a bug in the published version, or you have made a custom tweak to one of the published packs).

I can clarify this a little better and incorporate parts of your suggestion.

Reply:
Hmmm...thinking about this some more, it's even more complicated. You _can_ specify `--additional-packs` when you run `pack publish`. This means that you can publish packs using source resolution to override the default resolution. I thought we had disabled that, but we didn't. I can't see a situation where this would be a good thing to do (since you will be publishing your pack with contents that don't match the qlpack.yml).

I'll make sure to mention that doing so is not recommended.

Reply:
To be clear, I was distinguishing source dependencies resolved from the same workspace (which 100% _should_ be in the published pack) with source dependencies resolved from `--additional-packs`, which probably shouldn't.

The reason is that with dependencies in the workspace, we expect that all packs are tightly connected and should change in lock-step with each other. Also, they should all be available in the same git repo, so it is easy to edit all of them at once.

Source dependencies from `--additional-packs` can come from anywhere and it may be hard to track down exactly which versions are being used.

However...it's also true that all published query packs contain all of their transitive dependencies, so even if you override some of the expected dependencies with source dependencies, you can inspect them in the published pack. 

Apologies for going into the weeds here. I'm trying to figure out what is important to mention and what is superfluous for the docs.

Reply:
@felicitymay I'm not sure if this paragraph answers your questions. 

https://github.com/github/codeql/pull/10182/commits/94a6506cbb6899c709a47edca0a2a5109d20dba4#diff-8c14f9abe9524e8201b042c7288e073c2db2fb3d7adb88a5bf79427879a018e9R49-R54

Please let me know if this is clear enough to you.

Reply:
Thanks. That looks clearer to me.